src/corehost/src/tpafile.cpp

```diff
@@ -57,195 +75,188 @@ bool read_field(pal::string_t line, int& offset, pal::string_t& value_recv)
-            if (file.length() > len)
+            if (file.length() <= ext.length())
             {
-                // Extract the same amount of text from the end of file name
-                auto file_ext = file.substr(file.length() - len, len);
-
-                // Check if this file name matches
-                if (pal::strcasecmp(ext, file_ext.c_str()) == 0)
-                {
-                    // Get the assembly name by stripping the extension
-                    // and add it to the set so we can de-dupe
-                    auto asm_name = file.substr(0, file.length() - len);
-
-                    // TODO(anurse): Also check if already in TPA file
-                    if (added_assemblies.find(asm_name) == added_assemblies.end())
-                    {
-                        added_assemblies.insert(asm_name);
-
-                        tpaentry_t entry;
-                        entry.asset_type = pal::string_t(_X("runtime"));
-                        entry.library_name = pal::string_t(asm_name);
-                        entry.library_version = pal::string_t(_X(""));
-
-                        pal::string_t relpath(dir);
-                        relpath.push_back(DIR_SEPARATOR);
-                        relpath.append(file);
-                        entry.relative_path = relpath;
-                        entry.asset_name = asm_name;
-
-                        trace::verbose(_X("adding %s to TPA list from %s"), asm_name.c_str(), relpath.c_str());
-                        m_entries.push_back(entry);
-                    }
-                }
+                continue;
             }
+
+            auto file_name = file.substr(0, file.length() - ext.length());
+            if (file_name.empty() || m_local_assemblies.count(file_name))
+            {
+                continue;
+            }
+
+            auto file_ext = file.substr(file_name.length());
+            if (pal::strcasecmp(ext.c_str(), file_ext.c_str()))
+            {
+                continue;
+            }
+
+            pal::string_t file_path = dir + DIR_SEPARATOR + file;            
+            trace::verbose(_X("adding %s to local list from %s"), file_name.c_str(), file_path.c_str());
+            m_local_assemblies.emplace(file_name, file_path);
         }
     }
 }
 
-void tpafile::write_tpa_list(pal::string_t& output)
+void add_tpa_asset(
+    const pal::string_t& asset_name,
+    const pal::string_t& asset_path,
+    std::set<pal::string_t>* items,
+    pal::string_t* output)
 {
+    if (items->count(asset_name))
+    {
+        return;
+    }
+
+    trace::verbose(_X("adding tpa entry: %s"), asset_path.c_str());
+    output->append(asset_path);
+    output->push_back(PATH_SEPARATOR);
+    items->insert(asset_name);
+}
+
+void tpafile::write_tpa_list(
+        const pal::string_t& app_dir,
+        const pal::string_t& package_dir,
+        const pal::string_t& clr_dir,
+        pal::string_t& output)
+{
+    get_local_assemblies(app_dir);
```

Comment:
I, now, see ow you are using get_local_assemblies. Based upon that, I think it should be called something like "InitializeLocalAssemblyList".